rcl_yaml_param_parser/include/rcl_yaml_param_parser/parser.h

```diff
@@ -0,0 +1,64 @@
+// Copyright 2018 Apex.AI, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCL_YAML_PARAM_PARSER__PARSER_H_
+#define RCL_YAML_PARAM_PARSER__PARSER_H_
+
+#include <stdlib.h>
+
+#include "rcl_yaml_param_parser/types.h"
+#include "rcl_yaml_param_parser/visibility_control.h"
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+/// \brief Init param structure
+/// \param[in] allocator memory allocator to be used
+/// \return a pointer to param structure on success or NULL on failure
+RCL_YAML_PARAM_PARSER_PUBLIC
+rcl_params_t * rcl_yaml_node_struct_init(
+  const rcutils_allocator_t allocator);
+
+/// \brief Free param structure
+/// \param[in] params_st points to the populated paramter struct
+/// \param[in] allocator memory allocator to be used
+RCL_YAML_PARAM_PARSER_PUBLIC
+void rcl_yaml_node_struct_fini(
+  rcl_params_t * params_st,
+  const rcutils_allocator_t allocator);
```

Comment:
Agree. I think we can move the allocator field in rcl_variant_t into rcl_params_t and initialize it in rcl_yaml_node_struct_init(). Then remove "allocator" variable from other public API's except rcl_yaml_node_struct_init() and use the internal one. If you already have a fix can go ahead with it. If not I can make that change.

Reply:
Sounds perfect. I don't have the fix already.

Reply:
Thinking about it, because of the user specific state parameter in allocators, just passing the allocator in init() and cahing it in rcl_params_t and not able to pass the allocator in the fini() and other API's might defeat the purpose of the state field in the allocator. To overcome that, we can store the allocator in rcl_params_t, and let the user pass the allocator in other API's like fini and parse_yaml_file as it is currently implemented.. But, have to check if the incoming allocator is same as the stored allocator except the state value. Does it make sense?

Reply:
> Thinking about it, because of the user specific state parameter in allocators, just passing the allocator in init() and cahing it in rcl_params_t and not able to pass the allocator in the fini() and other API's might defeat the purpose of the state field in the allocator.

I think the expectation is that it's the allocator's state rather than a user's state. For example, someone might implement an allocator that only uses big chunk of memory at the start of the program. Calling `allocator.allocate(..., allocator.state)` would make the allocator record how much of that initial pool it's used. Since  `allocator.state` is a `void *` the state pointer itself must be the same when init, parse, and fini are called.

Reply:
OK, so are you suggesting, allow allocator to be passed in init, parse and fini and  check all the fields of the allocator in the beginning of the API of parse and fini to that  stored in rcl_params_t? If so, is there a API in rcutils or rcl which checks if two allocators are same or should I just implement one in parser.c to compare two allocators. I could not find a API in rcutils to compare  two allocators.

Reply:
> so are you suggesting, allow allocator to be passed in init, parse and fini and check all the fields of the allocator in the beginning of the API of parse and fini to that stored in rcl_params_t?

I recommend only allowing the allocator to be passed into init.

Reply:
Done!